[4-9단계 - 방탈출 예약 관리] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
[프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/4188)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

- [x] 1. **한 메서드에 오직 한 단계의 들여쓰기만 한다**
- [x] 2. **else 표현을 사용하지 않는다**
- [x] 3. **모든 원시 값과 문자열을 포장한다**
- [x] 4. **한 줄에 점을 하나만 사용한다**
- [x] 5. **이름을 줄여 쓰지 않는다(축약 금지)**
- [x] 6. **모든 엔티티를 작게 유지한다**
- [x] 7. **3개 이상의 스위프트 기본 데이터타입(Int, String, Double 등) 프로퍼티를 가진 타입을 구현하지 않는다**
- [x] 8. **일급 콜렉션을 사용한다**
- [ ] 9. **getter/setter를 구현하지 않는다**

- 8/9 = 88%  = 4

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

## 용어 정리
1. Repository = DAO ?
- DAO(DatabaseAccessObject) : DB에 접근하여 필요한 작업을 담당하는 객체
- Repository : 계층 구조의 최하위 영역으로, Database와 맞닿아 있는 영역.
- @Repository : 스프링에서 DAO를 정의할 때 붙이는 어노테이션. 해당 클래스를 스프링 빈으로 등록하며, DB 관련 예외를 스프링의 DataAccess 예외로 변환해줌.
- 결론 : 스프링에서 @Repository를 붙이기 때문에 관용적으로 Dao 클래스의 이름을 Repository로 짓는다. 의미 상으로 명확히 구분하려면 Repository 패키지에 각 Dao가 속하는 것이 맞다고 생각하지만(Repository.ReservationDao) 관용적으로 클래스명도 Repository로 (Repository.ReservationRepository)사용하는 맥락도 이해할 수 있다.

## 레이어 분리에 대한 사고 흐름
8단계 미션에서 Reservation 객체와 ReservationTime 객체사이의 의존성을 처리해야 하는 상황이 생겼습니다.
1. (문제발생) ReservationTime의 Id로 ReservationTime 테이블에 접근해 ReservationTime 객체를 생성하는 것이 Reservation을 등록하기 위한 선제 조건이 되었습니다.
2. (해결) Reservation과 ReservationTime 각각의 Repository는 분리하되 1.을 위해 두 Repository를 동시에 활용하는 방법으로 상위 계층 레이어 생성을 선택했습니다.
3. 비즈니스 로직에 따라 각 Repository에게 적절한 DB 요청을 보내는 2. 의 레이어를 ReservationService 로 명명하였습니다.
4. (문제발생) ReservationRepository와 ReservationTimeRepository에 접근하는데 공통적으로 ReservationService를 사용합니다. -> ReservationTimeController에서 Time이 아닌 Reservation에 접근할 수 있게 되었습니다.
5. (해결) ReservationTimeService를 만들어 서비스 별로 각각의 관심사를 분리하였습니다. (TimeRepository는 TimeService를 통해서만 접근할 수 있게)
6. (결론) 가장 근본적으로 도메인 객체 간에 의존성이 있기 때문에 두 도메인을 동시에 다루는 현재의 비즈니스 로직에서 의존성을 완전히 분리할 수는 없습니다. 지금과 같이 Service Layer 간에 의존성을 두는 것이 최선이라고 생각합니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jye326 committed:
feat(controller) : 홈 화면 구현

- admin 요청

jye326 committed:
feat(controller) : 예약 조회 기능 구현

jye326 committed:
refactor(controller) : 예약 추가/취소 기능 리팩터링

- post : 생성자 오버로딩 활용

jye326 committed:
refactor(controller) : 패키지 이동 및, toEntity 사용

jye326 committed:
docs(README) : 리드미 업데이트

jye326 committed:
refactor(model) : Dto 사용

- 생성자 단일화
- Dto 사용

jye326 committed:
feat(dto) : Dto

jye326 committed:
feat(controller, service) : 비즈니스 로직 service 레이어 분리

jye326 committed:
refactor(controller) : 단일 생성자 @Autowired 제거

jye326 committed:
rename(model) : AddReservationDto

jye326 committed:
rename(model) : AddReservationDto

jye326 committed:
refactor(model) : Reservation toEntity 이동

- Dto로 위치 변경

jye326 committed:
refactor(controller) : ReservationService -> Reservations

jye326 committed:
refactor(controller) : index 위치 변경

- Reservations -> Controller

jye326 committed:
feat(repository) : ReservationRepository 추가

- 사단계 테스트 구현 (DB 환경설정)

jye326 committed:
feat(controller) : DB 데이터 조회 코드 구현

jye326 committed:
RED) test(controller) : DB 데이터 추가, 삭제 테스트 작성

jye326 committed:
GREEN) feat(controller, repository) : DB 데이터 추가, 삭제 기능 작성

jye326 committed:
delete(model) : Reservations 미사용

jye326 committed:
refactor(controller) : html 페이지 반환 controller 분리, RequestMapping

jye326 committed:
refactor(controller) : 주입 방식 변경

Autowired -> 생성자 주입

jye326 committed:
feat(schema) : reservationTime 스키마 추가

jye326 committed:
feat(repository) : TimeRepository 추가

jye326 committed:
feat(controller) : TimeController 추가

jye326 committed:
RED) test(controller) : time Post, Get, Delete 테스트 작성

jye326 committed:
feat(model) : Time

jye326 committed:
feat(controller) : TimeController 추가

jye326 committed:
refactor(repository) : 패키지 이동

jye326 committed:
feat(dto) : AddTimeDto

jye326 committed:
refactor(domain) : rename(Time ->ReservationTime)

jye326 committed:
refactor(dto) : rename(AddReservationDto -> ReservationRequestDto)

jye326 committed:
fix(schema) : time 이름 변경 오류 수정

jye326 committed:
fix(controller) : 예약 관리 페이지 변경

- admin/reservation-legacy -> admin/reservation

jye326 committed:
fix(schema) : reservation 테이블 수정

- 칼럼 수정 : time(varchar) -> time_id(bigint)
- 외래키 추가 : time_id - reservation_time (id)

jye326 committed:
refactor(dto) : ReservationTimeDto toEntity 제거

jye326 committed:
fix(model) : time 필드 타입 변경 -> ReservationTime

jye326 committed:
feat(service) : ReservationService 레이어 추가

jye326 committed:
refactor(dto) : ReservationTimeDto toEntity 제거

jye326 committed:
feat(test) : MissionStepTest

- 팔단계 테스트 추가
- 기존 테스트 수정(변경 API 명세 적용)

jye326 committed:
feat(test) : MissionStepTest

- 구단계 테스트 추가

jye326 committed:
merge conflict : step1-step2

jye326 committed:
refactor(all) : 코드 포맷팅, 람다식 적용

jye326 committed:
refactor(all) : 코드 포맷팅, 람다식 적용

theminjunchoi commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,13 @@
+import roomescape.model.Reservation;
+import roomescape.model.ReservationTime;
+
+public record ReservationRequestDto(@JsonProperty(value = "name", defaultValue = "name") String name,
```
저도 리뷰어분께 피드백 받은 내용인데, XXXRequest, XXXResponse라고만 명명을 해도, dto임을 명시할 수 있어서 클래스명이 길어질 경우에는 필요에 따라 dto를 안적어도 될 것 같아요

jye326 replied:
저는 아직은 XXXRequest만 보고 당연히 이 클래스는 Dto겠구나 인식할 정도로 해당 네이밍이 익숙하지는 않은 것 같습니다. 코드를 작성하는 저의 기준에 맞추어 네이밍을 하고자 하는 의도가 있었습니다. 
> 이름을 줄여쓰지 않는다

원칙도 지킬 겸.

sure-why-not replied:
> 이름을 줄여쓰지 않는다

해당 원칙은 축약어를 말하는 것이긴 합니다. ㅎㅎ 예) Information -> Info
Dto는 무엇을 의미하나요? 해당 record는 api요청 역할을 하고 있는것으로 보여서요.

jye326 replied:
HTTP 요청이라는 데이터를 프로그램에서 사용할 수 있는 객체로 변환하기 위해 만들어진 클래스라고 생각합니다! Dto라는 것을 굳이 명시하는 것이 그렇게 불편하지 않아서 수정하지 않았습니다!

theminjunchoi commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @GetMapping("/times")
+    public ResponseEntity<List<ReservationTime>> addTime() {
```
메서드명이 잘못된 것 같아요! getTime를 쓰고 싶었던 걸까요?

jye326 replied:
실수가 있었네요. 발견해주셔서 감사합니다!

theminjunchoi commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,43 @@
+    @GetMapping("admin/time")
+    public String adminTime() {
+        return "admin/time";
+    }
```
이 패키지안에 AdminPageController가 있으면 이 부분은 옮겨도 될 것 같아요!
추가로 이를 분리하면 ReservationTimeController는 `@RestController`, `@RequestMapping("/times")`로 어노테이션을 명시할 수 있을 것 같아요!

jye326 replied:
마침 딱 필요하다고 느낀 부분이네요!! 감사합니다!

theminjunchoi commented on src/main/java/roomescape/ReservationService.java:
```diff
@@ -0,0 +1,47 @@
+import roomescape.repository.ReservationTimeRepository;
+
+@Service
+public class ReservationService {
```
좀 더 책임분리가 가능할 것 같아요! 
RerservationTime과 관련된 책임들을 ReservationTimeService로 묶을 수 있지 않을까요? 혹시 일부러 이렇게 했다면 이유도 궁금합니다~

추가로 Request를 만들었는데 Response를 만들지 않은 이유도 궁금해요.
도메인 객체를 Controller에게 넘겨주면, 외부에서 비즈니스 로직을 건들 수 있는 위험이 있다고 생각이 듭니다!
또 Response라는 dto를 도입함으로써 얻을 수 있는 추가적인 이점도 생각해보면 좋을 것 같아요.

jye326 replied:
Service라는 레이어 자체의 필요성에 대한 고민이 있었고, 결국 ReservationTimeService로 분리했습니다. 자세한 과정은 PR Discription에 있습니다!

ResponseDto는 오로지 보안을 위해 존재한다고 생각합니다. 도메인 객체(ex. User)가 계층을 그대로 넘나들게 되면
1. 전송을 원하지 않는 민감한 필드(ex. User의 비밀번호)가 노출될 위험이 있다.
2. 의도치 않은 계층(위치)에서 도메인 객체의 메서드를 호출할 위험이 있다.
결국 두 가지 경우 모두 데이터 전송 시 의도치 않은 데이터가 포함 되었을 때의 위험 사항으로 볼 수 있습니다.

그런데 이번 미션에서는 적어도 지금까지는 각 도메인 모델에 메서드가 포함되지도 않고, 사용하지 않는 필드도 없습니다. 따라서 ResponseDto의 필요성이 낮다고 보아 구현 우선순위에서 밀렸습니다 ㅎㅎ (언젠가는 구현해야지~ 하고 인지는 하고 있답니다)

하지만 계층 간 데이터 전송 시 습관적으로 Dto를 사용하는 것도 유효하다고 생각합니다.  Dto가 미리 작성되어 있으면 계층 간 데이터 이동 시 위험에 대해 한 번 더 의식할 수 있을 테니까요!

theminjunchoi commented on src/main/java/roomescape/ReservationService.java:
```diff
@@ -0,0 +1,47 @@
+
+    public Integer deleteReservation(long id) {
+        return reservationRepository.deleteReservation(id);
+    }
```
이건 제 사견인데요, 사용자가 존재하지 않은 id의 예약기록을 지워주려고 했을 때, 적절한 에러를 던져주는 것도 중요한 것 같아요.
실제 여러 서비스들을 보면, 데이터를 삭제할 때 이미 존재하지 않는 경우에 피드백을 주는 경우가 많더라구요!

현재 코드를 보면, 존재하지 않은 id의 예약기록을 지워주려고 할 때 에러없이 코드가 실행되는 것 같아서 말을 꺼내봤습니다!

jye326 replied:
흠. 그럴 수도 있겠다 싶은 부분입니다. 존재하지 않는 Time의 Id를 기반으로 Reservation을 등록하려는 요청에 대한 처리도 같은 맥락이라고 생각하는데요.

변명 한 번 해보겠습니다.
미션에서 사용하는 웹 페이지를 확인해보면 사용자가 예약 등록을 요청하기 위해 존재하는 Time을 먼저 선택해야 함을 알 수 있습니다. 삭제도 마찬가지로 존재하는 예약을 확인하여 삭제 버튼을 눌러야 하죠. 그래서 안했습니다!😄 실제로 발생할 수 없는 상황이라고 생각했으니까요!

물론 실제로는 백엔드가 프론트 구조를 다 알 수 없으니 가능한 모든 상황에 대한 예외처리가 필요하다고 꼬집으시면 정말 할 말이 없습니다. 하하

theminjunchoi reviewed (commented):
안녕하세요 띠용~
코드 너무 잘봤습니다.
띠용이 오고가면서 했던 고민들이 반영된 코드라서 재밌게 읽을 수 있었어요!
많이 배워갑니다!

jye326 committed:
refactor(service, repository) : 서비스 레이어 추가

- ReservationTimeService 추가
- ReservationService 의존 관계 변경
 - ReservationRepository
 - ReservationTimeService

jye326 committed:
refactor(controller) : ReservationTimeController 어노테이션 수정

- page 반환 로직 이동
- REST Controller 명시

jye326 committed:
refactor(controller) : rename method

jye326 committed:
refactor(model) : 원시값 포장

- name -> UserName
- date -> ReservationDate
- ReservationDateTime 추가

jye326 committed:
feat(model) : Validate UserName

- 검증 기능 및 테스트 추가

jye326 committed:
feat(model) : Validate ReservationTime

- 검증 기능 및 테스트 추가

jye326 committed:
feat(model) : Validate ReservationDate

- 검증 기능 및 테스트 추가

jye326 committed:
feat(model) : Validate ReservationDateTime

- 검증 기능 및 테스트 추가

sure-why-not commented on src/main/java/roomescape/controller/AdminPageController.java:
```diff
@@ -0,0 +1,22 @@
+        return "admin/reservation";
+    }
+
+    @GetMapping("admin/time")
```
어드민 시간 관리 페이지에서 추가기능이 동작하지 않네요.

jye326 replied:
js파일 링크가 잘못되어 있었네요! 수정했습니다!

sure-why-not commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,40 @@
+        return ResponseEntity.ok(reservationService.addReservation(reservationRequestDto));
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Integer> deleteReservation(@PathVariable long id) {
+        return ResponseEntity.ok().body(reservationService.deleteReservation(id));
+    }
```
삭제 api에서 200응답과 id를 내려주는 이유는 무엇인가요~?

jye326 replied:
id를 내려준다는 것이 id를 반환한다는 의미일까요? jdbc update 함수는 해당 작업으로 영향을 받는 칼럼의 수를 반환하는 것으로 알고 있습니다! 현재도 id가 아닌 삭제된 칼럼의 수를 반환합니다! 200 응답은 서버에서 삭제 응답을 정상적으로 처리했음을 전달합니다!

sure-why-not replied:
> id를 내려준다는 것이 id를 반환한다는 의미일까요? 

아 삭제된 칼럼수를 반환하는 것을 잘못 말했네요. ㅎㅎ

> jdbc update 함수는 해당 작업으로 영향을 받는 칼럼의 수를 반환하는 것으로 알고 있습니다! 

jdbc update 함수와 HTTP API와는 어떤 연관이 있나요?
왜 삭제된 칼럼수를 응답으로 내려주나요?

jye326 replied:
음.. HTTP 응답 값에 대해서는 크게 고민하지 않았던 것 같습니다! 단순히 update가 반환하는 값을 나도 반환하면 될 것 같다고 생각했습니다!

생각해보니 delete 오류 발생 시 200번대 응답을 주지 않으며, 식별자인 id로 delete하는 경우, 성공시 무조건 1을 반환하므로 불필요한 반환값인 것 같아 반환값을 없앴습니다.

sure-why-not commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+public record ReservationRequestDto(@JsonProperty(value = "name", defaultValue = "name") String name,
+                                    @JsonProperty("date") String date,
+                                    @JsonProperty("timeId") Long timeId) {
+    public static Reservation toEntity(Long id, ReservationRequestDto reservationRequestDto, ReservationTime reservationTime) {
```
ReservationRequestDto 포함 모든 필드값을 파라미터로 받고있는데,  일반 메소드가 아닌 static 메소드로 구현한 이유는 무엇인가요?

jye326 replied:
entity 변환은 객체 없이도 수행할 수 있어야 할 것 같아 static으로 선언하였습니다만, ReservationRequestDto 자신을 파라미터로 받는 이상, static 이 오히려 불필요하게 느껴져 수정하였습니다!

불필요한 to 파라미터가 삭제되니 가독성도 좋아진 것 같습니다!

sure-why-not commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+
+public record ReservationRequestDto(@JsonProperty(value = "name", defaultValue = "name") String name,
+                                    @JsonProperty("date") String date,
+                                    @JsonProperty("timeId") Long timeId) {
```
snake case가 아닌 camel case로 정하신 이유가 있으신가요?

jye326 replied:
미션 테스트의 요청 파라미터 형식을 맞추고자 우선 camel로 작성하였습니다! 통일성을 위해 테스트와 함께 snake case로 변경하였습니다!

sure-why-not replied:
아직 camel case인 것 같아요~!

jye326 replied:
camel로 통일했습니다!

sure-why-not commented on src/main/java/roomescape/dto/ReservationTimeRequestDto.java:
```diff
@@ -0,0 +1,6 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record ReservationTimeRequestDto(@JsonProperty(value = "startAt") String startAt) {
```
시간정보를 String으로 받는 이유가 무엇인가요?

jye326 replied:
날짜 정보와 동일하게, 미션 테스트가 제공하는 파라미터 형태를 따라가고자 함과 LocalTime.parse 등의 호환이 가능함이 있습니다!

sure-why-not commented on src/main/java/roomescape/dto/ReservationTimeRequestDto.java:
```diff
@@ -0,0 +1,6 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record ReservationTimeRequestDto(@JsonProperty(value = "startAt") String startAt) {
```
`@JsonProperty`를 사용하는 이유도 궁금해요~

jye326 replied:
미션 초반 페어 프로그래밍 과정에서 `@JsonProperty`를 사용하지 않으면 컴파일러에 따라 `-parameter` 옵션 여부에 따른 역직렬화 오류가 있을 수 있다고 배워 사용하고 있습니다!

sure-why-not commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+import roomescape.model.UserName;
+
+public record ReservationRequestDto(@JsonProperty(value = "name", defaultValue = "name") String name,
+                                    @JsonProperty("date") String date,
```
날짜정보를 String으로 받는 이유는 무엇인가요~?

jye326 replied:
해당 부분은 크게 고민하지 않고 미션에서 제공된 뼈대 코드를 따라가고 있었습니다! 이후 LocalDateTime.parse를 활용하기도 편리한 형태인 것 같아 변경하지 않았습니다.

sure-why-not commented on src/main/java/roomescape/model/ReservationTime.java:
```diff
@@ -0,0 +1,23 @@
+
+public class ReservationTime {
+    private final Long id;
+    private final String start_at;
```
요건 왜 snake case죠?👀

jye326 replied:
데이터 베이스 스키마에 맞추어 전부 snake case로 통일하였습니다!

jye326 replied:
아닙니다. java코드는 camel로 통일했습니다.

sure-why-not commented on src/main/java/roomescape/repository/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,49 @@
+
+    public ReservationTime getReservationTimeById(Long id) {
+        String sql = "select start_at from reservation_time where id = ?";
+        String start_at = jdbcTemplate.queryForObject(sql, String.class, id);
```
1. 변수명을 snake case로 사용하는 이유가 있으신가요?
2. String 타입으로 받는 이유가 있나요? LocalTime 타입을 고민해보면 좋을 것 같아요~

jye326 replied:
1. DB 스키마와 맞추고 싶어 snake case로 작성했습니다. -> java 내에서는 Camel, DB와 sql에서는 snake를 사용하도록 수정했습니다!
2. LocalTime 매핑 관련 Jackson 설정을 공부한 후 적용해 보겠습니다!

sure-why-not replied:
자바 명명규칙은 카멜케이스에요~!

jye326 replied:
java 내에서는 Camel, DB와 sql에서는 snake를 사용하도록 수정했습니다!

sure-why-not commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,55 @@
+                    sql, new String[]{"id"});
+            ps.setString(1, reservationRequestDto.name());
+            ps.setString(2, reservationRequestDto.date());
+            ps.setLong(3, reservationRequestDto.timeId());
```
매번 인자를 순서를 고려하여 넣어줘야 하는데요.
JdbcTemplate의 다른 구현체를 적용해보시면 순서에 얽매이지 않고 사용할 수 있어요. 🙂

jye326 replied:
다른 구현체가 있는 줄 몰랐습니다! `NamedParameterJdbcTemplate`에 대해 공부해보고 적용하겠습니다!

sure-why-not replied:
학습해보고 적용해보시면 좋을 것 같아요~

sure-why-not commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public List<Reservation> getAllReservations() {
+        String sql = "SELECT r.id, r.name, r.date, r.time_id, t.start_at FROM reservation as r inner join reservation_time as t on r.time_id = t.id";
```
select * from ~ 을 하는 경우도 있는데 어떤 기준이 있나요?

jye326 replied:
join을 사용했기 때문에 *을 사용하면 어떤 table에서 온 필드인지 알 수 없어 각각 명시하였습니다!

sure-why-not commented on src/main/java/roomescape/model/ReservationDateTime.java:
```diff
@@ -0,0 +1,35 @@
+import java.time.LocalTime;
+
+public class ReservationDateTime {
+    private final ReservationDate date;
+    private final ReservationTime time;
```
값객체를 고려한 점은 좋습니다.
java에서 제공하는 LocalDate 타입을 고려해보면 검증하는 로직과 파싱하는 로직을 제거할 수 있겠네요.

jye326 replied:
요청 파라미터가 String이라 LocalDate나 LocalTime으로 변경하면 역직렬화가 불가능할까 싶어 변경하지 않았습니다. 
그런데 막상 변경해보니 예상 외로 역직렬화에 별다른 문제가 없었습니다! 어찌된 일인가 싶어 알아보니

JSON 데이터를 변환하는 Jackson 라이브러리에 `jackson-datatype-jsr310`(Java Time 모듈)이 등록되어 있고, 이 모듈 안에 `LocalDateDeserializer`가 기본 ISO-8601 포맷(yyyy-MM-dd)의 문자열을 LocalDate로 파싱해 주기 때문이라는 것을 깨달았습니다.

> 1. JSON에 "date": "2025-08-05"처럼 ISO 형식 문자열이 들어오면
> 2. Jackson이 자동으로 LocalDate.parse("2025-08-05")를 호출해서
> 3. Reservation.date 필드(LocalDate 타입)에 매핑

sure-why-not commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -68,4 +95,118 @@ public class MissionStepTest {
     }
+
+    @Test
+    void 사단계() {
```
<img width="339" alt="스크린샷 2025-04-26 오후 2 00 38" src="https://github.com/user-attachments/assets/1b0a2c61-db8b-4d85-930b-da15a3a2f6bb" />

테스트가 깨지는데 확인 부탁드려요~

jye326 replied:
![image](https://github.com/user-attachments/assets/153e8fb6-2237-4305-ae2f-c44b19b1effc)

확인하였습니다!

sure-why-not reviewed (changes requested):
띠용 안녕하세요! ☺️
4-9단계 미션 하느라 고생 많으셨습니다.
몇가지 코멘트 남겨두었으니 확인 부탁드려요~!

jye326 committed:
refactor(util) : util 패키지 이동

jye326 committed:
fix(resource) : js 파일 연결

jye326 committed:
fix(resource) : 출발 시간 필드명 수정

jye326 committed:
fix(resource) : js 파일 수정

- Reservation 객체 필드값 포장 적용

jye326 committed:
fix(test) : 과거 예약 불가 테스트 적용

jye326 committed:
fix(test) : dto 변경

- timeId, startAt 변수 스타일 변경 camel -> snake
- toEntity static 제거

jye326 committed:
feat(repository) : 리포지토리 인터페이스 생성

jye326 committed:
feat(repository) : 메모리 레포지토리 생성

jye326 committed:
feat(repository) : 메모리 레포지토리 생성

jye326 committed:
feat(service) : 리포지토리 인터페이스 적용

jye326 committed:
refactor(model) : Reservation 필드명 변경

- name:UserName -> userName:UserName
- 필드명 불일치로 인한 Jackson 매핑 오류 해결

jye326 committed:
test(service) : ReservationServiceTest 작성

- 메모리 리포지토리 사용

jye326 committed:
refactor(model) : 필드명 스타일 변경

- camel Case 적용 : ReservationTime.start_at -> startAt, ReservationTimeRequestDto.time_id -> timeId

jye326 committed:
refactor(model) : 필드명 스타일 변경

- camel Case 적용 : ReservationTime.start_at -> startAt, ReservationTimeRequestDto.time_id -> timeId

jye326 committed:
fix(model) : change ReservationTime.startAt type

- String -> LocalTime

jye326 committed:
fix(model) : change ReservationDate.date type

- String -> LocalDate

m-a-king commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @GetMapping()
+    public ResponseEntity<List<Reservation>> reservations() {
```
메서드 명은 동사가 좋지 않을까요?

sure-why-not replied:
코멘트 남겨주신대로 자바 메서드명은 동사로 시작합니다~

jye326 replied:
반영하였습니다!

m-a-king commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @PostMapping()
+    public ResponseEntity<Reservation> addReservation(@RequestBody ReservationRequestDto reservationRequestDto) {
```
add 라고만 적는다면 가독성이 떨어질까요?

jye326 replied:
ReservationController니까 add만 보고도 addReservation을 유추할 수 있음을 말씀하시는걸까요?
저는 둘 다 괜찮다고 보는데 지금의 네이밍도 가독성이 떨어진다고 보이지 않아서 유지하겠습니다!

m-a-king commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Integer> deleteTime(@PathVariable Long id) {
+        return ResponseEntity.ok(reservationTimeService.deleteTime(id));
```
삭제된 행의 수를 알려주는 이유가 뭘까요?

jye326 replied:
별 다른 이유는 없었고 jdbc.update에서 반환하는 값을 그대로 반환하였습니다. 그런데 생각해보니 delete 오류 발생 시 200번대 응답을 주지 않으며, 식별자인 id로 delete하는 경우, 성공시 무조건 1을 반환하므로 불필요한 반환값인 것 같아 반환값을 없앴습니다.

m-a-king commented on src/main/java/roomescape/model/ReservationDateTime.java:
```diff
@@ -0,0 +1,35 @@
+        LocalDateTime now = LocalDateTime.now();
+        if (dateTime.isBefore(now)) {
+            throw new DateTimeException("과거 예약은 불가능합니다.");
+        }
```
인정합니다.

m-a-king commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final ReservationTimeRepository reservationTimeRepository;
+
+    public ReservationTimeService(ReservationTimeRepository jdbcReservationTimeRepository) {
```
인정합니다.

m-a-king commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE reservation_time
+(
+    id   BIGINT       NOT NULL AUTO_INCREMENT,
+    start_at VARCHAR(255) NOT NULL,
```
맞습니다.

m-a-king reviewed (approved):
😄

jye326 committed:
fix(model, service) : 과거 예약 점검 시점 변경

- 과거 예약 요청은 불가
- 과거 예약에 대한 조회 및 객체 생성은 가능

jye326 committed:
fix(model, service) : 과거 예약 점검 시점 변경

- 과거 예약 요청은 불가
- 과거 예약에 대한 조회 및 객체 생성은 가능

jye326 committed:
Merge remote-tracking branch 'origin/step2' into step2

# Conflicts:
#	src/test/java/roomescape/service/ReservationServiceTest.java

sure-why-not commented on src/main/java/roomescape/repository/MemoryReservationTimeRepository.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public ReservationTime addTime(LocalTime start_at) {
```
자바 언어 기본 명명규칙은 카멜케이스입니다. ㅎㅎ
전체적으로 snake case로 적용된 부분 확인 부탁드려요~

sure-why-not reviewed (changes requested):
띠용 안녕하세요!
리뷰 반영 전체적으로 잘 해주셨네요. 👍🏻
몇가지 추가 코멘트 남겨두었으니 확인 부탁드려요~!

jye326 committed:
refactor(controller) : rename function

- reservations -> getReservations

jye326 committed:
fix(controller) : change delete response type

- Integer -> Void

jye326 reviewed (commented):
중간에 커밋이 한 번 꼬였습니다... 고쳤다고 해 놓고 안 고쳐진 부분이 있었다면 죄송합니다..

jye326 committed:
fix(app) : change code style

- snake case -> camel case

sure-why-not reviewed (approved):
띠용 안녕하세요! 🙂
리뷰 반영 잘 해주셨네요. 👍🏻
이번 미션은 여기서 머지해도 될 것 같아요.
다음 미션도 화이팅이에요~~!!💪🏻
